### DISCUSSION TITLE
Projected user org member permissions

### DISCUSSION THREAD

**Discussion by muhlemmer** [zitadel/zitadel#7256] [Ideas]
Currently ZITADEL defines organization and instance member roles and permissions in `defaults.yaml`. The permission check is done on API call level. For example: "is this user allowed to make this call on this org". This makes sense on the V1 API where the API is permission-level shaped. For example, a search for users always happens in the context of the organization. (Either the organization the calling user belongs to, or through member ship and the `x-zitadel-orgid` header.

## The problem

However, the v2 API is resource based. For example the "users" resource. From a resource perspective organizations will be a different resource and the fact that users belong to organizations is somewhat irrelevant. When an API client is calling the v2 users API to search users, it should be able to search all users from their instance. The fact the users belong to an organization does no longer matter. The caller can still search for user from a particular org, but this is not a hard requirement.

> The v2 user search query can only return users from organizations where the Calling User has the permission `user.read `

- Permission `user.read` must be a constant defined by the search query. As the query "knows" what resources it is going to read.
- The Calling User is a variable to the query.

So a primary filter has to be done:

> Find all organization where `user_id` has permission `user.read`

However, the current model does not allow:

- Search organizations by `user_id` and `roles`.  `roles` is an array of text in the `projections.org_members` table.
- Search `roles` by `permission`, because permissions are defined as array values to a role key in `defaults.yaml`.
- `permissions` are not in the database at all.

This gives a new challenge to managing permissions. The calling user can have `user.read` permission on 0 or more organizations. Using the current permission-model will need to search the database for all memberships a user has and which roles that user has on each organization. Then, search the permissions from runtime configuration (`defaults.yaml`) and boil it down to a list of organizations the user is to read and apply it as a `resource_owner` filter on the user query. This is a huge hassle and probably bad performance due to the extra round trips to the DB.

## Proposal

Move permissions into the database. This is an example schema loosely based on current projections, roles and permissions but trimmed down for demonstration purposes. It creates the possibility to search organizations by `user_id` of the Calling User and the required permission.

```sql
create schema projections;

create table projections.users
(
  id text not null,
  creation_date timestamptz not null,
  resource_owner text not null,
  instance_id text not null,
  primary key(instance_id, id)
);

create table projections.role_permissions
(
  role text not null,
  permission text not null,
  
  primary key(role, permission)
);

create table projections.org_user_roles (
  org_id text not null,
  user_id text not null,
  role text not null,

  primary key(org_id, user_id, role)
);

create view projections.org_user_permissions as
select r.org_id, r.user_id, p.permission 
  from projections.org_user_roles r
  join projections.role_permissions p on r.role = p.role;
```

<details><summary>Example data</summary>

```sql
insert into projections.users
  (id, resource_owner, instance_id, creation_date)
values
('1', 'org1', 'inst1', '2024-01-01 20:34:43+00'),
('2', 'org1', 'inst1', '2024-01-02 20:34:43+00'),
('3', 'org1', 'inst1', '2024-01-03 20:34:43+00'),
('4', 'org1', 'inst1', '2024-01-04 20:34:43+00'),
('5', 'org1', 'inst1', '2024-01-05 20:34:43+00'),
('6', 'org1', 'inst1', '2024-01-06 20:34:43+00'),
('7', 'org1', 'inst1', '2024-01-07 20:34:43+00'),
('8', 'org1', 'inst1', '2024-01-08 20:34:43+00'),
('9', 'org2', 'inst1', '2024-01-09 20:34:43+00'),
('10', 'org2', 'inst1', '2024-01-10 20:34:43+00'),
('11', 'org2', 'inst1', '2024-01-11 20:34:43+00'),
('12', 'org2', 'inst1', '2024-01-12 20:34:43+00'),
('13', 'org2', 'inst1', '2024-01-13 20:34:43+00'),
('14', 'org2', 'inst1', '2024-01-14 20:34:43+00'),
('15', 'org2', 'inst1', '2024-01-15 20:34:43+00'),
('16', 'org2', 'inst1', '2024-01-16 20:34:43+00'),
('17', 'org3', 'inst2', '2024-01-17 20:34:43+00'),
('18', 'org3', 'inst2', '2024-01-18 20:34:43+00');

insert into projections.role_permissions
(role, permission)
values
('ORG_USER_MANAGER', 'org.read'),
('ORG_USER_MANAGER', 'user.read'),
('ORG_USER_MANAGER', 'user.global.read'),
('ORG_USER_MANAGER', 'user.write'),
('ORG_USER_MANAGER', 'user.delete'),
('ORG_OWNER_VIEWER', 'org.read'),
('ORG_OWNER_VIEWER', 'user.read'),
('ORG_OWNER_VIEWER', 'user.global.read'),
('ORG_SETTINGS_MANAGER', 'org.read'),
('ORG_SETTINGS_MANAGER', 'org.write'),
('ORG_SETTINGS_MANAGER', 'org.member.read');

insert into projections.org_user_roles
(org_id, user_id, role)
values
('org1', '1', 'ORG_USER_MANAGER'),
('org2', '1', 'ORG_OWNER_VIEWER'),
('org1', '2', 'ORG_OWNER_VIEWER'),
('org2', '2', 'ORG_SETTINGS_MANAGER'),
('org2', '3', 'ORG_OWNER_VIEWER'),
('org2', '4', 'ORG_SETTINGS_MANAGER');
```

</details>

We can now build a search query that looks like:

```sql
select u.*
  from projections.users u
  
  -- permission join & filter
  join projections.org_user_permissions p on u.resource_owner = p.org_id
  where u.instance_id = $1
  and p.user_id = $2 and p.permission = 'user.read'
  -- search filters after
  and u.creation_date > $3
;
```

The query takes `instance_id` and the `user_id` of the user doing the search query.
It searches on which orgs the `user_id` has permission 'user.read' and joins that with the result set.
In this example we are searching for users created after a certain date, the third argument.

Try it out on [db fiddle.](https://dbfiddle.uk/QtdYTPZ-)

## Still to be defined

1. Write model. How do we push and reduce permissions into projections?
2. Do we allow users to change or add permissions?
3. Do we need similar model for instance level permissions, or can we project it directly to org level?

CC: @livio-a (auth related), @adlerhurst (db related), @stebenz (resource related)

**Comment by hifabienne** [comment]
When we have a look at this, we might also take into consideration, that admins are able to define their own roles and add the permissions. This is already defined in the following issue: https://github.com/zitadel/zitadel/issues/5257

**Comment by adlerhurst** [comment]
What might be worth is defining some use cases / requirements especially for queries to these tables. 

some that come to my mind:

* is admin allowed to change user x
* is user x allowed to change project y

This brings up the question if we should have a look at [zanzibar](https://research.google/pubs/zanzibar-googles-consistent-global-authorization-system/) at least for internal permission checks. Zanzibar is built to answer those questions, maybe more fine grained that we need it:

<img width="958" alt="image" src="https://github.com/zitadel/zitadel/assets/27845747/0dabc4e4-49ce-45b9-811e-d4e57bf9cbe2">

**Reply by muhlemmer** [reply]
Isn't that architecture a bit overkill for solving "User Alice is granted `user.read` on org 123"?

![image](https://github.com/user-attachments/assets/45c0a046-a790-4c0c-8f51-295711685a7c)
